### ISSUE TITLE
Possibility to set readPreference on a query-basis

### ISSUE THREAD

**Issue by balintant** [mongodb/laravel-mongodb#1555]
We can already specify the default `readPreference` for Moloquent, which is a nice thing to keep it on `preferSecondary` to lower the load on the primary node of the replicaSet.

In some cases though, you want to read from the primary:
- locking a ton of records and then fetching them - in order to ensure your processes never conflict
- or because of some reasons, you just need real up-to-date data... 

---

So far, I don't see any option to use `preferSecondary` by default, and overwrite that for specific queries (only). Please correct me if I'm wrong.

- Are we willing to have an easy-to-use solution for that as part of `laravel-mongodb`?
- For those of you who solved this issue, can you please share your experiences?

_Ps.: we know that this practice is not really recommended, but so far it seems to be the best option for us._

**Comment by lab08tonci** [comment]
Any update on this one?
@balintant, did you manage to get to a solution?

**Comment by balintant** [comment]
@lab08tonci unfortunately I left the project soon after I submitted this issue, and I didn't work with the package since that.

**Comment by sanketgandhi876** [comment]
@jenssegers Is there any way we can achieve this?

**Comment by jenssegers** [comment]
Does Laravel offer a way to switch connections on a model dynamically?

**Comment by sanketgandhi876** [comment]
@jenssegers It does offer `sticky` option. Please read this: https://laravel.com/docs/5.8/database#read-and-write-connections
My interpretation is when sticky option is enabled then in the same request cycle read query is executed using the same write connection. But when it is off then laravel should use write connection for write queries and read connection for reads.

**Comment by katjackson** [comment]
@jenssegers I would love to be able to use the aforementioned `sticky` option with mongodb. Any suggestions?

**Comment by katjackson** [comment]
@alcaeus thanks so much for your detailed explanation! I I agree that being able to set `readConcern`, `readPreference`, and `writeConcern` per operation would be the best solution, and indeed it is why this ticket was started. 

I understand that reading from the secondary is not their intended purpose, but we are attempting to identify specific processes in our system where this would be safe, in order to improve performance. I though that the `sticky` option might be a way to offload a little bit of the load without having to assess every process individually.

Since I cannot easily specify those settings per operation, my current plan is to set up a second connection to MongoDB where `readPreference` is `secondaryPrefered`. Then I can set that "read only" connection on the model before performing any queries that we know are safe, even if they fall victim to the <1% occurrence of replication lag.

**Comment by GromNaN** [comment]
Is your issue solved @katjackson?

**Comment by katjackson** [comment]
@GromNaN sort of? but it didn't work. 

the initial request of this ticket was to be able to set the readPreference on a per-operation basis. to achieve this, i set up 2 connections, one with `primaryPreferred` and one with `secondaryPreferred`. then, i added a method to a base model to switch the connection,
```
    public function setReadOnlyConnection()
    {
        $this->connection = 'mongodb_read_only';
        return $this;
    }
```
and used it for read operations that could withstand a few seconds lag.
```
        $client = (new Client())
            ->setReadOnlyConnection()
            ->where('key', $clientKey)
            ->project(['name' => 1])
            ->first();
```

But... the replication lag turned out to be much greater than expected (even though the `writeConcern` was always set to `majority`). On a day where we experienced significant performance issues on our MongoDB cluster, the secondary was over 45 min behind for most of the day. And, in the end, we decided that no queries are really safe to read from the secondary. Not at our scale anyway.
